lib/ace/ext/papeeria/latex_behaviour.coffee

```diff
@@ -0,0 +1,226 @@
+  # inside a comment, just on a start of it. That's the reason for `column != 0` check
+  isCommentToken = (token, column) -> column != 0 and token? and token.type.indexOf(COMMENT_TYPE) > -1
+
+
+  isEscapedInsertion = (token, column) -> (
+      token? and
+      (
+          token.type.indexOf(ESCAPE_TYPE) > -1 or
+          token.type.indexOf(KEYWORD_TYPE) > -1 or
+          token.type.indexOf(STORAGE_TYPE) > -1
+      ) and
+      column - token.start == 1
+  )
+
+
+  isInEquation = (session, row, column) ->
+    token = session.getTokenAt(row, column)
+    nextToken = session.getTokenAt(row, column + 1)
+    state = session.getState(row)
+    pState = session.getState(row - 1)
+    lastState = if typeof(state) == "string" then state else state[state.length - 1]
+    lastPrevState = if typeof(pState) == "string" then pState else pState[pState.length - 1]
+    return (
+        # This handles the case, when the cursor is on the empty string
+        (not token? and lastPrevState.indexOf(MATH_TYPE) > -1) or
+        # This handles the common case, when the cursor is inside the equation
+        (token? and token.type.indexOf(EQUATION_TYPE) > -1) or
+        # This handles the specific case, when the cursor is on the very start of the equation
+        (nextToken? and nextToken.type.indexOf(EQUATION_TYPE) > -1)
+    )
+
+
+  DO_NOTHING = null
+  AUTO_INSERT = { text: "$$", selection: [1, 1] }
+  SKIP = { text: "", selection: [1, 1] }
+  dollarsInsertionAction = (state, action, editor, session, text) ->
+    if text == '$' and not editor.inMultiSelectMode
+
+      { row, column } = editor.getCursorPosition()
+      line = session.getLine(row)
+
+      selection = editor.getSelectionRange()
+      selected = session.getTextRange(selection)
+      # If some text is selected, we surround it with $
+      if selected != ""
+        if editor.getWrapBehavioursEnabled()
+          return getWrapped(selection, selected, text, text)
+        else
+          return DO_NOTHING
+
+      token = session.getTokenAt(row, column)
+      nextToken = session.getTokenAt(row, column + 1)
+
+      # If cursor is inside a comment or escaped, do nothing
+      if isCommentToken(token, column) or isEscapedInsertion(token, column)
+        return DO_NOTHING
+
+      prevChar = line[column - 1] or ''
+      nextChar = line[column] or ''
+
+      # If cursor is in equation, either skip closing $ or do nothing
+      if isInEquation(session, row, column)
+        return if nextChar == '$' then SKIP else DO_NOTHING
+
+      # Otherwise, insert or skip
+      shouldSkip = (nextChar == '$' and (prevChar != '$' or nextToken.type.indexOf(RPAREN_TYPE) > -1))
+      return if shouldSkip then SKIP else AUTO_INSERT
+
+
+  dollarsDeletionAction = (state, action, editor, session, range) ->
+    if editor.inMultiSelectMode
+      return DO_NOTHING
+
+    selected = session.doc.getTextRange(range)
+    if range.isMultiLine() or selected != '$'
```

Comment:
maybe just check `selected == '$'` ? If it is `$` then it is automatically not multiline, no?